akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -423,7 +433,9 @@ private[http] object HttpServerBluePrint {
         def onPull(): Unit =
           if (oneHundredContinueResponsePending) pullSuppressed = true
           else if (!hasBeenPulled(requestParsingIn)) pull(requestParsingIn)
-        override def onDownstreamFinish() = cancel(requestParsingIn)
+        override def onDownstreamFinish(): Unit =
+          if (openRequests.isEmpty) completeStage()
+          else failStage(new IllegalStateException("User handler flow was cancelled with ongoing request") with NoStackTrace)
```

Comment:
This new behavior is more aggressive than before: if the user handler cancels the request stream, we will kill the connection. Would be nice to be able to get rid of that change as it's somewhat of an invitation of new race conditions. In practice it wouldn't probably not be so serious as user-handlers usually do not complete explicitly.